## Practice Question

In [14]:
with open("data_file.txt",encoding = "utf-8") as file:
    content = file.read()
    print(content)

Name: John D. Smith | Email: john.smith@example.com | Phone: (123) 456-7890 | Date: 2024-03-15 | IP: 192.168.1.1
Name: Emily-Rose Clark | Email: emily.c@mydomain.net | Phone: 123.456.7890 | Date: 2023/12/01 | IP: 10.0.0.2
Name: Ahmed Z | Email: az@hotmail.com | Phone: +1-800-555-0199 | Date: 15-Mar-2024 | IP: 172.16.0.5
Name:  王小明 | Email: ming.wang@china.cn | Phone: 0086-10-65551234 | Date: 2024.06.03 | IP: 192.0.2.10
Name: María González | Email: maria_g@example.org | Phone: (555)555-5555 | Date: 01.01.2024 | IP: 203.0.113.8
Name: Robert\nInvalid Email | Email: robert[at]example[dot]com | Phone: None | Date: unknown | IP: 999.999.999.999
Name: Alice Wonderland | Email: alice@wonder.land.uk | Phone: 555 123 4567 | Date: 2025-06-01 | IP: 8.8.8.8


##  Emails
* Extract all valid email addresses (hint: ignore the robert[at]example[dot]com one).

* Match only .com email addresses.

In [5]:
import re 
with open("data_file.txt",encoding = "utf-8") as file:
    content = file.read()
#all valid Emails
email_pattern = re.compile(r"([a-zA-Z0-9_.+%-]+)@([a-zA-Z0-9-]+)\.[a-zA-Z.]{2,}")

com_email_pattern = re.compile(r"([a-zA-Z0-9_.+]+)@([a-zA-Z0-9-]+)\.com") # only .com emails
all_emails = email_pattern.finditer(content)

print("All Valid Emails")

for match in all_emails:
    print(match.group())
print()
print("only .com Emails")
com_emails = com_email_pattern.finditer(content)
for match in com_emails:
    print(match.group())    


All Valid Emails
john.smith@example.com
emily.c@mydomain.net
az@hotmail.com
ming.wang@china.cn
maria_g@example.org
alice@wonder.land.uk

only .com Emails
john.smith@example.com
az@hotmail.com


##  Phone Numbers

* Extract all phone numbers in any format:
 +1-800-555-0199, (555)555-5555, 123.456.7890, etc.

* Extract only international phone numbers (start with + or 00). 

In [87]:
import re 
with open("data_file.txt",encoding = "utf-8") as file:
    content = file.read()
    

phone_number = re.compile(r"(\+?\d{1,3}[-.\s])?\(?\d{2,4}\)?[-.\s]?\d{3,4}[-.\s]?\d{4}")    

internetional_number = re.compile(r"(\+|00)\d{1,3}[-.\s]?\d{1,4}[-.\s]?\d{3,4}[-.\s]?\d{4}")

number =phone_number.finditer(content)
print("All Phone Numbers")
for match in number:
    print(match.group())
print()
number2 = internetional_number.finditer(content)
print("International Numbers Only")
for match in number2:
    print(match.group())    

All Phone Numbers
(123) 456-7890
123.456.7890
+1-800-555-0199
086-10-65551234
(555)555-5555
555 123 4567

International Numbers Only
+1-800-555-0199
0086-10-65551234


## Names

* Extract all full names (between Name: and | Email:).


In [62]:
import re 
with open("data_file.txt",encoding = "utf-8") as file:
    content = file.read()

full_name = re.compile(r"Name:\s(.+?)\s*\| Email") 

matches = full_name.finditer(content)
for match in matches:
    print("Name: ",match.group(1))  # by group(1) we only get the name 

Name:  John D. Smith
Name:  Emily-Rose Clark
Name:  Ahmed Z
Name:   王小明
Name:  María González
Name:  Robert\nInvalid Email
Name:  Alice Wonderland


## Dates

* Extract all dates regardless of format: YYYY-MM-DD, DD-MMM-YYYY, YYYY/MM/DD, etc. 

In [2]:
import re 
with open("data_file.txt",encoding = "utf-8") as file:
    content = file.read()
date_pattern = re.compile(r"Date:\s*\d{2,4}[-/.](\d){2,4}[-./]\d{2,4}")

#date_pattern = re.compile(r"Date:\s*\d{2,4}[-/.]?[a-zA-Z0-9]{3}+[-./]\d{2,4}") #this give this format Date: 15-Mar-2024
matches = date_pattern.finditer(content)
for match in matches:
    print(match.group())    

Date: 2024-03-15
Date: 2023/12/01
Date: 2024.06.03
Date: 01.01.2024
Date: 2025-06-01
